akka-stream/src/main/scala/akka/stream/impl/Timers.scala

```diff
@@ -138,13 +138,21 @@ object Timers {
 
   }
 
-  final class IdleTimeoutBidi[I, O](val timeout: FiniteDuration) extends GraphStage[BidiShape[I, I, O, O]] {
+  /**
+   * @param timeoutToTermination allows customising the behaviour once a timeout triggers.
+   *                             Returning None causes completion, and Some(exception) causes failure of the stage with that exception.
+   */
+  final class IdleTimeoutBidi[I, O](val timeout: FiniteDuration, timeoutToTermination: () ⇒ Option[Throwable]) extends GraphStage[BidiShape[I, I, O, O]] {
```

Comment:
> I don't want completion. I want specific exception types thanks to which I know if it was a tcp idle timeout or some random timeout which I don't care about. This would solve akka/akka-http#651 because I could match on the specific exception and give a proper error: https://github.com/akka/akka-http/pull/650/files#diff-623ef36775342f2db6b86855f7ffea64R418 hinting to people that they can change the specific setting to tune this behaviour.

`source.recover(TimeoutException => throw myFunnyException)`

> I need a way of implementing this, without causing Akka Stream TCP to log errors and panic.

Logs at DEBUG level. 

> Because most of the time this timeout is fine and no one should get errors from that.

Then why failing the stream, why not closing it simply?

> recover would log on ERROR if I throw in it, which is why we'd need the mapError, because I don't want this class to log anything. I will do the logging, but I need a specific exception, not a random TimeoutException to give users a good experience.

Then implement mapError. There is no need to add ad hoc options to the timer stages.